### PR TITLE
test: a test framework for verifying upgrade of Zoe and ZCF

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
refs: #6678

## Description

a test framework for verifying upgrade of Zoe and ZCF. This includes bootstrap scripts that can be used to upgrade Zoe and ZCF.

The test starts from the currently installed version of Zoe and ZCF. First it verifies that reallocation via staging and via the helper work and that the version internal to ZCF is not present. It then upgrades Zoe and ZCF as necessary to introduce the new behavior. Finally, it re-runs the initial verification to show that the ZCF internal version works.

This currently fails on the first step since it runs in a state in which the new Zoe/ZCF code has already replaced the old. To check it in, we will change the parity of the test that the new internal `zcf.atomicRearrange()` is not present, since this version runs in an environment where that change has already been made. 

Variants of this test will run on master without this PR to verify that upgrading to the new Zoe/ZCF works, and in
deployment/upgrade-test to show that the upgrade to the new zoe & zcf will succeed. A copy of this code run on master without this PR, with bundles containing the new Zoe & ZCF code will show that the upgrade from the current state succeeds, and zcf's behavior in the probe contract will be different before and after. 

### Security Considerations

Validates the upgrade plan.

### Scaling Considerations

None

### Documentation Considerations

None

### Testing Considerations

It's all about testing.

### CODE REVIEW

File: packages/vats/test/bootstrapTests/supports.js [hunk 1]
```diff
@@ -488,4 +497,94 @@ export const makeSwingsetTestKit = async (
     timer,
   };
 };
+
 /** @typedef {Awaited<ReturnType<typeof makeSwingsetTestKit>>} SwingsetTestKit */
+
+export const makeTestContext = async t => {
```
Comment (turadg) [inline packages/vats/test/bootstrapTests/supports.js]: it's only the context for `test-vats-restart`. why move this from that module to this shared one?

File: packages/vats/test/bootstrapTests/supports.js [hunk 2]
```diff
@@ -17,14 +18,22 @@ import { loadSwingsetConfigFile } from '@agoric/swingset-vat';
...
+  makeAgoricNamesRemotesFromFakeStorage,
   slotToBoardRemote,
 } from '../../tools/board-utils.js';
+import { makeWalletFactoryDriver, makeZoeDriver } from './drivers.js';
 
 // to retain for ESlint, used by typedef
 E;
 
+// main/production config doesn't have initialPrice, upon which 'open vaults' depends
+const PLATFORM_CONFIG = '@agoric/vats/decentral-itest-vaults-config.json';
```
Comment (turadg) [inline packages/vats/test/bootstrapTests/supports.js]: I don't think this should be in specified in `supports`.  Each test can choose.

File: packages/vats/test/bootstrapTests/supports.js [hunk 3]
```diff
@@ -488,4 +497,94 @@ export const makeSwingsetTestKit = async (
...
+  return {
+    ...swingsetTestKit,
+    agoricNamesRemotes,
+    walletFactoryDriver,
+    buildProposal,
+  };
+};
+
+export const makeZoeTestContext = async t => {
+  console.time('DefaultTestContext');
```
Comment (turadg) [inline packages/vats/test/bootstrapTests/supports.js]: consider ZoeTestContext. but also, why put in the common supports? It's used only one in test.

File: packages/vats/test/bootstrapTests/test-zcf-upgrade.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    /** @type {ERef<import('../../src/types.js').BridgeHandler>} */
+    const coreEvalBridgeHandler = await EV.vat('bootstrap').consumeItem(
+      'coreEvalBridgeHandler',
+    );
+    await EV(coreEvalBridgeHandler).fromBridge(bridgeMessage);
+  };
+  const source = `${dirname}/${ZCF_PROBE_SRC}`;
+
+  const zcfProbeBundle = await bundleSource(source);
+  // uncomment and add `import fs from "fs";` to generate a bundle of the prober contract
```
Comment (turadg) [inline packages/vats/test/bootstrapTests/test-zcf-upgrade.js]: please clarify under what conditions someone would need to do this

**Review by turadg — CHANGES_REQUESTED** [review]
Please keep the context setup in the modules that use them. "supports" is meant for more generic code. If we ever need multiple bootstrap tests that use the same context setup we can re-evaluate.

File: packages/vats/test/bootstrapTests/supports.js (hunk 3 above)
Comment (Chris-Hibbert) [reply packages/vats/test/bootstrapTests/supports.js]: changed, and moved

File: packages/vats/test/bootstrapTests/supports.js (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/vats/test/bootstrapTests/supports.js]: I thought there would be more sharing.  I moved it back.

File: packages/vats/test/bootstrapTests/test-zcf-upgrade.js (hunk 4 above)
Comment (Chris-Hibbert) [reply packages/vats/test/bootstrapTests/test-zcf-upgrade.js]: It's only needed to build the prober which will be used in the deployment test. Once it has been built once, and checked in with #8018, it won't need to be done again.

But the reviewer of that PR may reasonably demand the ability to reproduce the process, so we describe it here.

File: packages/vats/test/bootstrapTests/supports.js (hunk 2 above)
Comment (Chris-Hibbert) [reply packages/vats/test/bootstrapTests/supports.js]: moved with makeTestContext()

**Review by turadg — APPROVED** [review]

**Pull request merged**